[Spring] Limit spring force to some axis

alxbilger opened the pull request:
This work is initially from @Camille-K in his branch https://github.com/Camille-K/sofa/tree/dev_limited_RestShapeSpringsForceField


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

Commit:
[ForceField] Add direction option in RestShapeSpringsFF

The idea of this commit is to add the ability for a user to select in
which directions the RestShapeSpringsFF will be effective during
simulation. For this we add a new data field containing 6 booleans: one
for each direction of effort (3 directions of translation, 3 of
rotation). For each entry, "true" indicates that the direction will be
taken into account, and "false" that it won't. The default value of the
parameter is set to [1 1 1 1 1 1], so that all directions are taken into
account if the parameter is not specifically provided by the user. This
way, we expect the component not to be breaking.

/!\ At the moment, we only implemented the restriction for translations,
but not for rotations. Concretely, changing one of the last 3 entries of
the data field to "false" doesn't have any effect on the simulation. The
implementation for rotations should be the object of further commits.

Commit:
[ForceField] Add test scene to test the new RestShapeSpringsFF option

Commit:
[ForceField] Extend direction option in RestShapeSpringsFF

Following previous commit, we extend the direction option to rotations.

Commit:
[ForceField] Update RestShapeSpringsFF test scene

Commit:
[ForceField] Fix warning in test scene

Commit:
[ForceField] Add callback in RestShapeSpringForceField

The callback is meant to recompute the internal variables of the
components if the input MechanicalObject indices are changed
dynamically.

alxbilger commented:
[ci-build][with-all-tests]

alxbilger committed:
Make activeDirections more generic

alxbilger committed:
No need to be a Vec

alxbilger committed:
make description matches dimension

alxbilger commented:
@adagolodjo Could you just remind us the goal of these changes?

alxbilger committed:
Fix wrong size

alxbilger committed:
Set component state

alxbilger commented:
One thing I had to do compared to @Camille-K 's changes was to set the component state to valid. Otherwise, the condition on the validity of the component state (also introduced in this PR) prevents to run the force field properly. I am curious if it worked in @Camille-K work.

alxbilger committed:
Fix scene

damienmarchal commented:
Hello @alxbilger 

Nice PR. 

This is outside of the scope of the PR but the use of the very intrusive
```cpp
if (this->d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
        return;
```
makes me thing one day we will have to do something. The way to go is probably to make clear when the component state validation/update issupposed to happen. The two choice are: "everytime/everywhere", update all the time including "in-between" timestep or only once at beginning of timestep (with visitor in the BeginAnimationStep event ?).

adagolodjo commented:
> 

Hello @alxbilger, I apologize for the delayed response. The objective is to provide the option to apply the RestshapSpring exclusively to the predefined axes, rather than applying it to all three or six degrees of freedom as we currently do.

alxbilger committed:
Remove check of component state

This is going to be checked earlier (but in another pull request)

alxbilger committed:
Refactor setting the component state

It now follow the same pattern used in other components: the component state is set to valid only at the end
